common4j/src/main/com/microsoft/identity/common/java/authorities/NativeAuthCIAMAuthority.kt

```diff
@@ -0,0 +1,112 @@
+//  AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+//  LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+//  OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+//  THE SOFTWARE.
+
+package com.microsoft.identity.common.java.authorities
+
+import com.microsoft.identity.common.java.exception.ClientException
+import com.microsoft.identity.common.java.logging.LogSession
+import com.microsoft.identity.common.java.logging.Logger
+import com.microsoft.identity.common.java.providers.nativeauth.NativeAuthOAuth2Configuration
+import com.microsoft.identity.common.java.providers.nativeauth.NativeAuthOAuth2Strategy
+import com.microsoft.identity.common.java.providers.nativeauth.NativeAuthOAuth2StrategyFactory
+import com.microsoft.identity.common.java.providers.oauth2.OAuth2StrategyParameters
+
+// TODO risk: this Authority class is not composable through Authority.getAuthorityFromAuthorityUrl()
+//  which is the method that's used throughout the project to create Authorities.
+class NativeAuthCIAMAuthority (
+    private val authorityUrl: String,
+    val clientId: String
+) : CIAMAuthority(authorityUrl) {
+    companion object {
+        private val TAG = NativeAuthCIAMAuthority::class.java.simpleName
+        private val NATIVE_AUTH_USE_OPENID_CONFIGURATION = false
+
+        @Throws(Exception::class)
+        fun getAuthorityFromAuthorityUrl(authorityUrl: String, clientId: String):
+                NativeAuthCIAMAuthority {
+            // Piggy back on the existing authority creation to improve reliability.
+            val authority = Authority.getAuthorityFromAuthorityUrl(authorityUrl)
+
+            // Check what authority was returned (must be CIAM or Native Auth CIAM)
+            return if (authority !is CIAMAuthority) {
+                throw ClientException(ClientException.NATIVE_AUTH_INVALID_CIAM_AUTHORITY)
+            } else if (authority is CIAMAuthority && authority !is NativeAuthCIAMAuthority) {
+                // Authority returned was a base CIAM authority
+                NativeAuthCIAMAuthority(
+                    authorityUrl = authority.authorityUri.toString(),
+                    clientId = clientId
+                )
+            } else {
+                // Authority is already a NativeAuthCIAMAuthority
+                authority as NativeAuthCIAMAuthority
+            }
+        }
+    }
+
+
+    // TODO audience, slice, flight parameters, multiple clouds supported,
+    // isAuthorityHostValidationEnabled (AzureActiveDirectoryOAuth2Configuration). Consider extending
+    // AzureActiveDirectoryAuthority
+    init {
+        mAuthorityTypeString = "AAD_NA" // AAD Native Auth
+        mAuthorityUrlString = authorityUrl
+    }
+
+    private fun createNativeAuthOAuth2Configuration(challengeTypes: List<String>?): NativeAuthOAuth2Configuration {
+       LogSession.logMethodCall(TAG, "${TAG}.createNativeAuthOAuth2Configuration")
```

Comment:
Do we really need to log every single method call? Also what's the log level on these?

Reply:
We do, as for private preview we have no telemetry. We need to depend fully on local logs to do debugging and customer support.
What is your objection against it?

Reply:
The log level is info.

Reply:
A public api call can result in method execution for hundreds of internal methods. Some of those methods many be doing miner things such as StringUtil methods. Do we need to know about each method call in our logs? If the logs contain too many such statements then they may not be so useful.

Reply:
"Do we need to know about each method call in our logs?" - we don't, we're not calling `LogSession.logMethodCall` on every method that's called, only the ones we're interested in. We're tracking these as breadcrumbs, because for our private preview we have no telemetry and thus very limited insight into what's happening in our SDK. 
"If the logs contain too many such statements then they may not be so useful." - perhaps. we plan to clean this up once we get some real experience with this and get some actual bugs and logs in. There are probably improvements that can be made. Are there any objections against having it like this for the first version?

Reply:
I think this is fine as long as it is used in moderation.